lib/Sema/SemaBounds.cpp

```diff
@@ -1707,62 +1707,173 @@ namespace {
+          ProofResult UpperBoundsResult = CompareUpperOffsets(R, Cause, EquivExprs);
+
+          if (LowerBoundsResult == ProofResult::True &&
+              UpperBoundsResult == ProofResult::True)
+            return ProofResult::True;
+          if (LowerBoundsResult == ProofResult::False ||
+              UpperBoundsResult == ProofResult::False)
+            return ProofResult::False;
+        }
+        return ProofResult::Maybe;
+      }
+
+      // This function compares two lower offsets and returns true in the following cases:
+      // If both lower offsets are constant,
+      //   - regardless of their signs, if both of them evaluate to zero, return true.
+      //   - otherwise, the constant values will be compared if their have matching signs.
+      // If both lower offsets are non-constant, return true if EqualValue() returns true.
+      // If R's lower offset is an unsigned expression and this lower offset is zero, return true.
+      ProofResult CompareLowerOffsets(BaseRange &R, ProofFailure &Cause, EquivExprSets *EquivExprs) {
+        if (IsLowerOffsetConstant() && R.IsLowerOffsetConstant()) {
+          if (LowerOffsetConstant.getExtValue() == 0 &&
+              R.LowerOffsetConstant.getExtValue() == 0)
+            return ProofResult::True;
+          if (SignsMatch(UpperOffsetConstant, R.UpperOffsetConstant) &&
+              LowerOffsetConstant <= R.LowerOffsetConstant)
+            return ProofResult::True;
+          Cause = CombineFailures(Cause, ProofFailure::LowerBound);
+          return ProofResult::False;
+        }
+        if (IsLowerOffsetVariable() && R.IsLowerOffsetVariable() &&
+            EqualValue(S.Context, LowerOffsetVariable, R.LowerOffsetVariable, EquivExprs))
+          return ProofResult::True;
+        if (R.IsLowerOffsetVariable() && IsLowerOffsetConstant() &&
+            R.LowerOffsetVariable->getType()->isUnsignedIntegerType() && LowerOffsetConstant.getExtValue() == 0)
+          return ProofResult::True;
+
+        return ProofResult::Maybe;
+      }
+
+      // This function compares two upper offsets and returns true in the following cases:
+      // If both uppper offsets are constant,
+      //   - regardless of their signs, if both of them evaluate to zero, return true.
+      //   - otherwise, the constant values will be compared if their have matching signs.
+      // If both upper offsets are non-constant, return true if EqualValue() returns true.
+      // If this upper offset is an unsigned expression and R's upper offset is zero, return true.
+      ProofResult CompareUpperOffsets(BaseRange &R, ProofFailure &Cause, EquivExprSets *EquivExprs) {
+        if (IsUpperOffsetConstant() && R.IsUpperOffsetConstant()) {
+          if (UpperOffsetConstant.getExtValue() == 0 &&
+              R.UpperOffsetConstant.getExtValue() == 0)
+            return ProofResult::True;
+          if (SignsMatch(UpperOffsetConstant, R.UpperOffsetConstant) &&
```

Comment:
I don't think that you need to check that the signs match for the integer constants.  We always represent integers using signed integers.